sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CountingSource.java

```diff
@@ -391,41 +383,18 @@ public boolean advance() throws IOException {
       if (Long.MAX_VALUE - source.stride < current) {
         return false;
       }
-      if (getSplitBacklogElements() <= 0) {
+      long nextValue = current + source.stride;
+      if (timeToEmit(nextValue).isAfter(Instant.now())) {
         return false;
       }
-      current += source.stride;
+      current = nextValue;
       currentTimestamp = source.timestampFn.apply(current);
-      totalProduced++;
       return true;
     }
 
-    /**
-     * Gets the size of the split backlog in number of elements.
-     */
-    private long getSplitBacklogElements() {
-      if (source.period.getMillis() == 0) {
-        return Long.MAX_VALUE;
-      }
-      long msElapsed = Instant.now().getMillis() - started.getMillis();
-      long expectedOutputs =
-          LongMath.checkedMultiply(source.elementsPerPeriod, msElapsed) / source.period.getMillis();
-      return expectedOutputs - totalProduced;
-    }
-
-    @Override
-    public long getSplitBacklogBytes() {
-      long backlogElems = getSplitBacklogElements();
-      if (backlogElems == Long.MAX_VALUE) {
-        return BACKLOG_UNKNOWN;
-      }
-      // 8 == Long.BYTES
-      long backlogBytes = backlogElems * 8;
-      // overflow protection
-      if (backlogBytes < backlogElems) {
-        return Long.MAX_VALUE;
-      }
-      return backlogBytes;
+    private Instant timeToEmit(long value) {
+      long periodForValue = value / source.elementsPerPeriod;
+      return firstStarted.plus(source.period.multipliedBy(periodForValue));
```

Comment:
this numeric rounding seems terrible if we have 100 elements per 100s, for example.